File: leetcode/Tree/leetcode-0106/106.go
package tree

// TreeNode Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(inorder []int, postorder []int) *TreeNode {
	if inorder == nil || len(inorder) == 0 {
		return nil
	}
	root := &TreeNode{postorder[len(postorder)-1], nil, nil}
	left := 0
	for i := 0; i < len(inorder); i++ {
		if inorder[i] == postorder[len(postorder)-1] {
			break
		}
		left++
	}
	root.Left = buildTree(inorder[:left], postorder[:left])
	root.Right = buildTree(inorder[left+1:], postorder[left:len(postorder)-1])
	return root
}

File: leetcode/Tree/leetcode-0106/106.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder == null || inorder.length == 0) {
            return null;
        }
        return getRoot(inorder, 0, inorder.length - 1, postorder, 0, postorder.length - 1);
    }

    public TreeNode getRoot(int[] inorder, int inStart, int inEnd, int[] postorder, int postStart, int postEnd) {
        if (inStart > inEnd) {
            return null;
        }

        TreeNode root = new TreeNode(postorder[postEnd]);
        int left = 0;
        for (int i = inStart; i <= inEnd; i++) {
            if (inorder[i] == postorder[postEnd]) {
                break;
            }
            left++;
        }
        root.left = getRoot(inorder, inStart, inStart + left - 1, postorder, postStart, postStart + left - 1);
        root.right = getRoot(inorder, inStart + left + 1, inEnd, postorder, postStart + left, postEnd - 1);
        return root;
    }
}